File: backend/app/schemas/voice.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.models.voice_profile import VoiceProfile, VoiceProfileStatus
from app.models.voice_sample import VoiceSample, VoiceSampleStatus


class CreateVoiceProfileRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    consent_confirmed: bool

    @field_validator("consent_confirmed")
    @classmethod
    def must_confirm_consent(cls, value: bool) -> bool:
        if not value:
            raise ValueError(
                "You must confirm you have permission to create and use this voice profile."
            )
        return value


class UpdateVoiceProfileRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)


class VoiceProfileResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    status: VoiceProfileStatus
    consent_confirmed: bool
    consent_confirmed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    sample_count: int
    valid_sample_count: int
    total_duration_seconds: float


class VoiceSampleResponse(BaseModel):
    id: uuid.UUID
    voice_profile_id: uuid.UUID
    original_filename: str
    mime_type: str
    file_size: int
    duration_seconds: float | None
    sample_rate: int | None
    channels: int | None
    status: VoiceSampleStatus
    validation_error: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CompleteVoiceProfileResponse(BaseModel):
    id: uuid.UUID
    status: VoiceProfileStatus
    valid_sample_count: int
    message: str


class ConversionResponse(BaseModel):
    conversion_id: uuid.UUID
    voice_profile_id: uuid.UUID
    status: str
    source_duration_seconds: float
    processing_time_seconds: float
    rtf: float
    model: str | None
    device: str
    output_url: str


class AudioValidationResultSchema(BaseModel):
    valid: bool
    duration_seconds: float | None
    sample_rate: int | None
    channels: int | None
    errors: list[str]


def build_voice_profile_response(
    profile: VoiceProfile, samples: list[VoiceSample]
) -> VoiceProfileResponse:
    valid_samples = [s for s in samples if s.status is VoiceSampleStatus.VALID]
    return VoiceProfileResponse(
        id=profile.id,
        name=profile.name,
        description=profile.description,
        status=profile.status,
        consent_confirmed=profile.consent_confirmed,
        consent_confirmed_at=profile.consent_confirmed_at,
        created_at=profile.created_at,
        updated_at=profile.updated_at,
        sample_count=len(samples),
        valid_sample_count=len(valid_samples),
        total_duration_seconds=sum(s.duration_seconds or 0.0 for s in valid_samples),
    )

File: frontend/types/voice.ts
export type VoiceProfileStatus =
  | "DRAFT"
  | "RECORDING"
  | "PROCESSING"
  | "READY_FOR_AI_PROCESSING"
  | "FAILED"
  | "ARCHIVED";

export type VoiceSampleStatus = "UPLOADED" | "VALIDATING" | "VALID" | "INVALID" | "DELETED";

export interface VoiceProfile {
  id: string;
  name: string;
  description: string | null;
  status: VoiceProfileStatus;
  consent_confirmed: boolean;
  consent_confirmed_at: string | null;
  created_at: string;
  updated_at: string;
  sample_count: number;
  valid_sample_count: number;
  total_duration_seconds: number;
}

export interface VoiceSample {
  id: string;
  voice_profile_id: string;
  original_filename: string;
  mime_type: string;
  file_size: number;
  duration_seconds: number | null;
  sample_rate: number | null;
  channels: number | null;
  status: VoiceSampleStatus;
  validation_error: string | null;
  created_at: string;
  updated_at: string;
}

export interface CompleteVoiceProfileResponse {
  id: string;
  status: VoiceProfileStatus;
  valid_sample_count: number;
  message: string;
}

export interface ConversionResult {
  conversion_id: string;
  voice_profile_id: string;
  status: string;
  source_duration_seconds: number;
  processing_time_seconds: number;
  rtf: number;
  model: string | null;
  device: string;
  output_url: string;
}
